### PR TITLE
chore(tests): added security tests

### PR DESCRIPTION
**Pull request by laurentlp**
This is one of many PRs that adds security tests to our HTTP API

Closes DEV-2057

### CODE REVIEW

File: packages/server/src/api.ts [hunk 1]
```diff
@@ -45,8 +45,14 @@ export class Api {
...
     this.root.use('/api', this.router)
-    this.router.use(cors())
-    this.router.use(express.json())
+    this.router.use(
+      cors({
+        origin: '*',
+        methods: ['GET', 'POST', 'DELETE'],
+        allowedHeaders: ['Content-Type', 'Authorization', 'Content-Length']
+      })
+    )
```
Comment (samuelmasse) [inline packages/server/src/api.ts:54]: What does this do?

File: packages/server/src/index.ts [hunk 2]
```diff
@@ -17,6 +17,7 @@ const launch = async () => {
   await setupEnv()
 
   const router = express()
+  router.disable('x-powered-by')
```
Comment (samuelmasse) [inline packages/server/src/index.ts:20]: Why this change? Is this related to security?

File: packages/server/src/sync/api.ts [hunk 3]
```diff
@@ -19,13 +20,18 @@ export class SyncApi {
     this.force = !!yn(process.env.SPINNED)
     this.schema = makeSyncRequestSchema(this.channels.list())
 
-    router.post('/sync', auth.public.auth(this.sync.bind(this)))
+    router
+      .route('/sync')
+      .post(auth.public.auth(this.sync.bind(this)))
+      .all(methodNotAllowed('POST'))
```
Comment (samuelmasse) [inline packages/server/src/sync/api.ts:26]: If we're only going to implement methodNotAllowed for this route we might as well just not bother with it. It's not really additional security anyways.

File: packages/server/test/security/api.test.ts [hunk 4]
```diff
@@ -0,0 +1,1005 @@
...
+
+        clients.first.userId = res.id
+
+        {
+          const res = await user(clients.second.clientId, clients.second.clientToken)
+
+          expect(res.id).not.toEqual(clients.first.userId)
+
+          clients.second.userId = res.id
+        }
```
Comment (samuelmasse) [inline packages/server/test/security/api.test.ts:268]: Why is there a random bracket here?

**Review by samuelmasse — APPROVED** [review]
Nice work! I have a few comments to clarify but overall these tests should prevent us from accidentally opening some of our routes to attacks.

I have a suggestion for another test : we could try sending payload of incorrect format, like url-econded

File: packages/server/src/api.ts (hunk 1 above)
Comment (laurentlp) [reply packages/server/src/api.ts:54]: This simply strengthens the CORS for our API. It basically tells that allowed requests from any origin, that we only allow GET, POST, and DELETE, and some headers. This is mostly for requests coming from a browser but it still good practice to be as restrictive as possible.

File: packages/server/src/index.ts (hunk 2 above)
Comment (laurentlp) [reply packages/server/src/index.ts:20]: Yes and no. It only removes the `x-powered-by: 'Express'` headers so that most people wouldn't necessarily know what framework powers our API. I know that we are an open-source project but again, it's only because it's a good practice thing. In our case it doesn't really have any negative nor positive effects.

File: packages/server/test/security/api.test.ts (hunk 4 above)
Comment (laurentlp) [reply packages/server/test/security/api.test.ts:268]: To create another block of code. That way I can redefine a res constant without having to name it res2 or something

File: packages/server/src/sync/api.ts (hunk 3 above)
Comment (laurentlp) [reply packages/server/src/sync/api.ts:26]: Again as a good practice, we should implement this everywhere. Best scenario, it narrows the attack vector on our endpoints, and worst case, we have a middleware that brings no overhead!
